packages/protocol/protocol.go

```diff
@@ -45,268 +43,520 @@ type Protocol struct {
 		}
-	}))
-
-	p.Events.Engine.NotarizationManager.EpochCommitted.Attach(event.NewClosure(func(details *notarization.EpochCommittedDetails) {
+	})
+	event.AttachWithWorkerPool(p.Events.CongestionControl.Scheduler.BlockScheduled, func(block *scheduler.Block) {
+		p.TipManager.AddTip(block)
+	}, wp)
+	event.AttachWithWorkerPool(p.Events.Engine.EvictionState.EpochEvicted, func(index epoch.Index) {
+		p.TipManager.EvictTSCCache(index)
+	}, wp)
+	event.AttachWithWorkerPool(p.Events.Engine.Consensus.BlockGadget.BlockAccepted, func(block *blockgadget.Block) {
+		p.TipManager.RemoveStrongParents(block.ModelsBlock)
+	}, wp)
+	event.AttachWithWorkerPool(p.Events.Engine.NotarizationManager.EpochCommitted, func(details *notarization.EpochCommittedDetails) {
 		p.TipManager.PromoteFutureTips(details.Commitment)
-	}))
-
-	p.Events.Engine.EvictionState.EpochEvicted.Attach(event.NewClosure(func(index epoch.Index) {
+	}, wp)
+	event.AttachWithWorkerPool(p.Events.Engine.EvictionState.EpochEvicted, func(index epoch.Index) {
 		p.TipManager.Evict(index)
-	}))
+	}, wp)
+}
 
-	p.Events.TipManager = p.TipManager.Events
+func (p *Protocol) onForkDetected(fork *chainmanager.Fork) {
+	claimedWeight := fork.Commitment.CumulativeWeight()
+	mainChainCommitment, err := p.Engine().Storage.Commitments.Load(fork.Commitment.Index())
+	if err != nil {
+		p.Events.Error.Trigger(errors.Errorf("failed to load commitment for main chain tip at index %d", fork.Commitment.Index()))
+		return
+	}
+
+	mainChainWeight := mainChainCommitment.CumulativeWeight()
+
+	if claimedWeight <= mainChainWeight {
+		// TODO: ban source?
+		p.Events.Error.Trigger(errors.Errorf("dot not process fork with %d CW <= than main chain %d CW received from %s", claimedWeight, mainChainWeight, fork.Source))
+		return
+	}
+
+	p.networkProtocol.RequestAttestations(fork.ForkingPoint, fork.EndEpoch(), fork.Source)
+}
+
+func (p *Protocol) switchEngines() {
+	p.activeEngineMutex.Lock()
+
+	if p.candidateEngine == nil {
+		p.activeEngineMutex.Unlock()
+		return
+	}
+
+	// Try to re-org the chain manager
+	if err := p.chainManager.SwitchMainChain(p.candidateEngine.Engine.Storage.Settings.LatestCommitment().ID()); err != nil {
+		p.activeEngineMutex.Unlock()
+		p.Events.Error.Trigger(errors.Wrap(err, "switching main chain failed"))
+		return
+	}
+
+	if err := p.engineManager.SetActiveInstance(p.candidateEngine); err != nil {
+		p.activeEngineMutex.Unlock()
+		p.Events.Error.Trigger(errors.Wrap(err, "error switching engines"))
+		return
+	}
+
+	// Stop current Scheduler
+	p.CongestionControl.Shutdown()
+
+	p.linkTo(p.mainEngine)
+
+	// Save a reference to the current main engine and storage so that we can shut it down and prune it after switching
+	oldEngine := p.mainEngine
+	oldEngine.Shutdown()
+
+	p.mainEngine = p.candidateEngine
+	p.candidateEngine = nil
+
+	p.activeEngineMutex.Unlock()
+
+	p.Events.MainEngineSwitched.Trigger(p.MainEngineInstance())
+
+	// TODO: copy over old epochs from the old engine to the new one
```

Comment:
What about this TODO?

Reply:
This is imho not really needed for the MVP, this is to allow node operators to keep long history even after chain switches. The majority of nodes will follow the main chain anyway and probably never do a chain switch.
This is something that we should plan and implement as a separate feature, since it will probably mean that we need to merge the chainmanager, enginemanager and databasemanager.